2021_RTCSA_dynfed/src/dynfed.rs

```diff
@@ -123,112 +123,135 @@ where
-    let dag_set_length = dag_set.len();
-    let mut dyn_fed_handlers = vec![DAGStateManager::new(); dag_set_length];
-    let mut ready_dag_queue: VecDeque<Graph<NodeData, i32>> = VecDeque::new();
-    let mut finished_dags_count = 0;
-
-    scheduler.set_processor(processor);
-
-    for (dag_id, dag) in dag_set.iter_mut().enumerate() {
-        dag.set_dag_id(dag_id);
-        dyn_fed_handlers[dag_id].set_minimum_cores_and_execution_order(dag, scheduler);
-    }
+pub struct DynamicFederatedScheduler<T>
+where
+    T: DAGSchedulerBase<HomogeneousProcessor>,
+{
+    pub dag_set: Vec<Graph<NodeData, i32>>,
+    pub processor: HomogeneousProcessor,
+    pub scheduler: T,
+}
 
-    let mut head_offsets: Vec<i32> = dag_set
-        .iter()
-        .map(|dag| dag.get_head_offset())
-        .collect::<HashSet<i32>>()
-        .into_iter()
-        .collect::<Vec<i32>>();
-    head_offsets.sort();
-    let mut head_offsets: VecDeque<i32> = head_offsets.into_iter().collect();
-
-    while finished_dags_count < dag_set_length {
-        if !head_offsets.is_empty() && *head_offsets.front().unwrap() == current_time {
-            dag_set
-                .iter_mut()
-                .filter(|dag| current_time == dag.get_head_offset())
-                .for_each(|dag| {
-                    ready_dag_queue.push_back(dag.clone());
-                });
-            head_offsets.pop_front();
+impl<T> DAGSetSchedulerBase<HomogeneousProcessor, T> for DynamicFederatedScheduler<T>
+where
+    T: DAGSchedulerBase<HomogeneousProcessor>,
+{
+    fn new(
+        dag_set: Vec<Graph<NodeData, i32>>,
+        processor: HomogeneousProcessor,
+        scheduler: T,
+    ) -> Self {
+        Self {
+            dag_set,
+            processor,
+            scheduler,
         }
+    }
 
-        //Start DAG if there are enough free core
-        while let Some(dag) = ready_dag_queue.front() {
-            let dag_id = dag.get_dag_id();
-            let num_processor_cores = processor.get_number_of_cores() as i32;
-            let total_allocated_cores = get_total_allocated_cores(&dyn_fed_handlers);
-
-            if dyn_fed_handlers[dag_id].can_start(num_processor_cores, total_allocated_cores) {
-                ready_dag_queue.pop_front();
-                dyn_fed_handlers[dag_id].start();
-            } else {
-                break;
-            }
+    fn schedule(&mut self) -> i32 {
+        let mut current_time = 0;
+        let dag_set_length = self.dag_set.len();
+        let mut dyn_fed_handlers = vec![DAGStateManager::new(); dag_set_length];
+        let mut ready_dag_queue: VecDeque<Graph<NodeData, i32>> = VecDeque::new();
+        let mut finished_dags_count = 0;
+
+        self.scheduler.set_processor(&self.processor);
+
+        for (dag_id, dag) in self.dag_set.iter_mut().enumerate() {
+            dag.set_dag_id(dag_id);
+            dyn_fed_handlers[dag_id]
+                .set_minimum_cores_and_execution_order(dag, &mut self.scheduler);
```

Comment:
By carving out this process into a separate function, the `scheduler` object can be removed from the argument of new method.